Adding documentation for code pipeline and workflow for rendering notifications in Calypso.

blackjackkent opened the pull request:
#### Changes proposed in this Pull Request

* Adds documentation at `apps/notifications/src/doc/note-rendering.md` giving a general overview of how notifications are processed to be rendered to users and the workflow and files necessary to modify the rendering logic.

Fixes #45439

kwight commented on apps/notifications/src/doc/note-rendering.md:
````diff
@@ -0,0 +1,105 @@
+yarn build
+```
+
+Then copy the built `dist` folder to your sandbox environment to test updated rendering behavior.
````
Can we make this easier with something like the `yarn dev --sync` command from PCYsg-ly5-p2?

kwight replied:
https://github.com/Automattic/wp-calypso/pull/45595

kwight commented:
This is great @blackjackkent 🚀  – can you add an example specific to a (Gutenberg) block maybe?

mmtr commented on apps/notifications/src/doc/note-rendering.md:
```diff
@@ -0,0 +1,105 @@
+
+## Troubleshooting and Updating Renderings
+
+Local development of notifications can take place making use of the Calypso dev server, as any changes made to `src/apps/notifications` will be reflected in the Calypso local development environment automatically.
```
Changes to the notifications endpoint might be also needed. 

In the particular example of a new post notification that includes blocks, this is the data returned by the API:

<img width="778" alt="Screen Shot 2020-09-10 at 13 58 50" src="https://user-images.githubusercontent.com/1233880/92726652-32a75280-f36e-11ea-9f4e-beb1586dc886.png">

Look at the `text` property. _Heading_ is the content of a `core/heading` block and _Paragraph_ is the content of a `core/paragraph` block. However, the API strips out the `h3` and `p` tags from them, so the components in `src/app/notifications` don't have any info about how to render them correctly [when converting the content to notification blocks](https://github.com/Automattic/wp-calypso/blob/8a700733fbc9564f4fa4d55454d7b40306d26f5c/apps/notifications/src/panel/templates/functions.jsx#L26).

mmtr replied:
Dug into this and found out that the post content is sanitized in `Notification_Block::add_post_block` when running `$block = $this->add_block( $post_content )` (source: `/wp-content/lib/class.notifications-builder.php:1046`).

Not sure if this should be documented here or if the FG page at PCYsg-46P-p2 is enough.

krymson24 replied:
I don't see what Miguel is seeing, but perhaps the endpoint that this data is coming from could be documented. This is what I see
<img width="885" alt="Screen Shot 2020-09-10 at 4 41 31 PM" src="https://user-images.githubusercontent.com/66652282/92802570-9f2d4c00-f384-11ea-99fc-000745d03fe8.png">

mmtr replied:
Have you tried expanding the `body` property?  It should be an array of objects that contain the `text` property I was referring to.

krymson24 replied:
Yeah -- I think I actually misunderstood your comment. I thought we should expect to see "Heading" and "Paragraph" specifically, but because we have different notifications, I didn't see those specifically. I'm realizing now you were just calling out how they're separated. 👍 
<img width="650" alt="Screen Shot 2020-09-11 at 3 48 06 PM" src="https://user-images.githubusercontent.com/66652282/92967060-98ffa400-f446-11ea-8273-8b566f1ae3ba.png">

mmtr commented:
So much helpful information on this readme! It pairs nicely with the FG documentation (PCYsg-g8e-p2) if we need more info (we can maybe add a link to this readme in the FG docs and vice versa).

Since the goal of https://github.com/Automattic/wp-calypso/issues/45439 is to offer additional dev resources for (Gutenberg) block developers, it'd be great to also add instructions on how to check how a (Gutenberg) block looks like in a notification. 

For me, the easiest way is to enable the "Notify new posts" check to a site in https://wordpress.com/following/manage:

<img width="325" alt="Screen Shot 2020-09-10 at 13 29 33" src="https://user-images.githubusercontent.com/1233880/92739752-81101d80-f37d-11ea-955f-640b3e5f9092.png">

Then publish a post containing the block to be checked.

krymson24 commented on apps/notifications/src/doc/note-rendering.md:
```diff
@@ -0,0 +1,105 @@
+# Notification Detail Rendering
+
+When a user clicks a note to view, that notification is rendered to the user in a slide-out drawer. The pipeline for rendering this drawer to the user begins in `/src/panel/templates/index.jsx`. When the user has selected a note to view, this component renders a child `Note` component from `src/panel/templates/note.jsx` with the property `detailView` set to `true`.
```
What's the subtlety between a "note" and "notification"?

blackjackkent replied:
They're the same thing -- both terms are used throughout the codebase and existing documentation. :/

krymson24 commented on apps/notifications/src/doc/note-rendering.md:
````diff
@@ -0,0 +1,105 @@
+After some initial processing, the `NoteBody` component performs a switch statement on each block object based on its type:
+
+```js
+switch ( block.signature.type ) {
````
Is adding the code here necessary? It may be sufficient to just refer to the source code. Adding the code here results in two places where this needs to be maintained

blackjackkent replied:
Good call -- I'll tweak that. :) Thanks.

krymson24 commented on apps/notifications/src/doc/note-rendering.md:
````diff
@@ -0,0 +1,105 @@
+}
+```
+
+Depending on the item's type, a different block component is rendered - `User` for a representation of a user (for instance, in a notification of a user liking a particular post), `Comment` for a comment made on a blog post, `Post` for a blog post (i.e. on a blog the user is following), `ReplyBlock` for a reply to a comment(?).
````
This seems self-documenting to me, but that's just how I see it

blackjackkent replied:
Eh, maybe a little, but I see some value in further clarifying (particularly for `User`, where the scenario in which a user block might come up doesn't seem entirely intuitive just from the name).

kwight replied:
> `ReplyBlock` for a reply to a comment(?)

Correct!

krymson24 commented on apps/notifications/src/doc/note-rendering.md:
```diff
@@ -0,0 +1,105 @@
+
+Depending on the item's type, a different block component is rendered - `User` for a representation of a user (for instance, in a notification of a user liking a particular post), `Comment` for a comment made on a blog post, `Post` for a blog post (i.e. on a blog the user is following), `ReplyBlock` for a reply to a comment(?).
+
+_n.b. - these 'blocks' are separate from the concept of 'blocks' in the Gutenberg block editor; a notification might have multiple blocks (for instance, multiple `User` blocks for multiple likes on a post), or it might have a single `Post` block, which then further renders Gutenberg blocks as part of the post content._
```
Interesting! Also, what's `n.b`?

blackjackkent replied:
Nota bene! https://en.wikipedia.org/wiki/Nota_bene

It's originally Latin for "note well", but basically shorthand for something like "as a side note". So in this case "As a side note, be aware that these are separate from Gutenberg blocks."

krymson24 commented on apps/notifications/src/doc/note-rendering.md:
````diff
@@ -0,0 +1,105 @@
+
+For instance, if the user has received a notification about a post on a blog they are following, the component in question will be `src/panel/notifications/block-post.jsx`:
+
+```js
````
Same file comment as https://github.com/Automattic/wp-calypso/pull/45529/files#r486603608

blackjackkent commented:
Pushed some updates based on the comments above. I'm keeping this documentation focused on what can be changed specifically within the Calypso codebase, but added links and reference to PCYsg-g8e-p2 for further information on notification architecture and PCYsg-46P-p2 for the WPCOM-related notification construction logic.

I also added a section giving more specific info on how to set up a Gutenberg post notification for testing post rendering and removed some of the code snippets that didn't need to be present (and would require maintaining as this code evolves).

Per discussion with @kwight , I will be setting up another PR working on adding an automated script to the notifications app for syncing local code to a sandbox for testing (similar to the `yarn dev --sync` script used in the full-site editing app - https://href.li/?PCYsg-ly5-p2 ). I'll also be doing some further digging into the specific fixes we're trying to accomplish for 217-gh-dotcom-manage to see if there's extra context there that would be useful for these docs. As both of these efforts may inform the content of this PR, I don't want to merge this yet, but further comments appreciated in the meantime.

krymson24 commented on apps/notifications/src/doc/note-rendering.md:
```diff
@@ -0,0 +1,105 @@
+
+Local development of notifications can take place making use of the Calypso dev server, as any changes made to `src/apps/notifications` will be reflected in the Calypso local development environment automatically.
+
+However, as notifications can also be viewed in an iframe on non-Calypso sites using the masterbar (see https://github.com/Automattic/wp-calypso/edit/master/apps/notifications/README.md), any changes to the notifications sub-application should also be tested on a sandboxed WP.com site. To do this, run the following commands:
```
What's the `masterbar`? Also I wonder if it's appropriate to add screenshots to READMEs. That's what I was planning to do.

kwight replied:
Our Jetpack/WP.com version of the logged-in admin bar:

<img width="956" alt="Screen Shot 2020-09-10 at 2 58 28 PM" src="https://user-images.githubusercontent.com/349751/92813499-1d362680-f376-11ea-8066-576ba5983fcf.png">

https://github.com/Automattic/jetpack/blob/865ea01b299b3926387f778de089cc4787602b73/modules/masterbar.php

kwight replied:
> Also I wonder if it's appropriate to add screenshots to READMEs

Sure, go for it – you could place them `/static/images`. We do avoid using huge amounts, really high resolution, or videos, to avoid bloating the repo's download size – in that case you could hang the image on a public service (like Cloudup), or commit it to WP.com.

kwight commented on apps/notifications/src/doc/note-rendering.md:
```diff
@@ -0,0 +1,131 @@
+
+If a block is rendering incorrectly due to formatting information being lost before the data reaches Calypso, you may need to perform further troubleshooting on the server-side logic used to generate the list of a user's notifications.
+
+This server logic can be found in the WPCOM codebase, and its primary entry point is in `/wp-content/lib/class.notifications-builder.php:1046` in a call to `Notification_Block::add_post_block()`. Further information about the server logic associated with notifications can be found in the PCYsg-46P-p2 FG entry.
```
Should this be another line number? L1046 is a `$statsd::timing()` call.

mmtr replied:
L1046 seems correct to me (I see a `$this->add_block()` call there) but yeah, we cannot point to a specific commit revision like in GitHub, so L1046 can get deprecated if new changes are pushed to `/wp-content/lib/class.notifications-builder.php`.

Maybe we can leave it as just `/wp-content/lib/class.notifications-builder.php`.

mmtr reviewed (approved):
Lovely. I think we can merge this and start identifying the involved pieces in the server/API (and document the result of that exploration in a P2 post / FG page).
